File: astify/token/Token.java
package astify.token;

import astify.core.Position;
import astify.core.Positioned;

public class Token implements Positioned {
    public final TokenType type;
    public final String value;
    public final Position position;

    public Token(TokenType type, String value, Position position) {
        this.type = type;
        this.value = value;
        this.position = position;
    }

    public boolean matches(TokenType type) {
        return this.type == type;
    }

    public boolean matches(TokenType type, String value) {
        return matches(type) && this.value.equals(value);
    }

    @Override public Position getPosition() {
        return position;
    }

    @Override public String toString() {
        if (type == TokenType.EOF) return "<EOF>";
        return type.toString() + " \"" + value.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n") + "\"";
    }
}

File: src/astify-runtime/astify/Token.kt
package astify

import java.util.*

abstract class Token(
        val position: TextPosition,
        val text: String
) {
    override fun equals(other: Any?): Boolean {
        if (other == null) return false
        if (other::class != this::class) return false
        if (other !is Token) return false
        return position == other.position && text == other.text
    }

    override fun hashCode() = Objects.hash(this::class, position, text)
    override fun toString() = "${this::class.simpleName}($text)"
}
